cmd/local-gadget/trace/dns.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/containerd/containerd/log"
+	"github.com/spf13/cobra"
+
+	commontrace "github.com/inspektor-gadget/inspektor-gadget/cmd/common/trace"
+	commonutils "github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/local-gadget/utils"
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	dnsTracer "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/tracer"
+	dnsTypes "github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/trace/dns/types"
+	localgadgetmanager "github.com/inspektor-gadget/inspektor-gadget/pkg/local-gadget-manager"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+const logalGadgetSubKey = "local-gadget-key"
+
+func newDNSCmd() *cobra.Command {
+	var commonFlags utils.CommonFlags
+
+	// The DNS gadget works in a different way than most gadgets: It
+	// attaches a new eBPF program to each container when it's
+	// created instead of using an eBPF map with the mount
+	// namespaces IDs to filter the events. For this reason we can't
+	// use the TraceGadget implementation here.
+	runCmd := func(*cobra.Command, []string) error {
+		localGadgetManager, err := localgadgetmanager.NewManager(commonFlags.RuntimeConfigs)
+		if err != nil {
+			return commonutils.WrapInErrManagerInit(err)
+		}
+		defer localGadgetManager.Close()
+
+		// local-gadget is designed to trace containers, hence enable this column
+		cols := dnsTypes.GetColumns()
+		col, _ := cols.GetColumn("container")
+		col.Visible = true
+
+		parser, err := commonutils.NewGadgetParserWithRuntimeInfo(&commonFlags.OutputConfig, cols)
+		if err != nil {
+			return commonutils.WrapInErrParserCreate(err)
+		}
+
+		eventCallback := func(event dnsTypes.Event) {
+			baseEvent := event.GetBaseEvent()
+			if baseEvent.Type != eventtypes.NORMAL {
+				commonutils.ManageSpecialEvent(baseEvent, commonFlags.Verbose)
+				return
+			}
+
+			// TODO: This enrich logic could be moved to the
+			// gadget itself to simplify this here and in
+			// pkg/gadget-collection/gadgets/trace/dns/gadget.go
+			container := localGadgetManager.GetContainer(event.Key)
```

Comment:
This was previously already implemented in pkg/gadget-collection/gadgets/trace/dns/gadget.go via `eventCallback()` calling `t.publishEvent(trace, &event, key)` without adding a Key in the event. Why is that PR removing it?

Reply:
As Jose mentioned, almost all the gadgets (tracers) enrich the data, but this one doesn't. I think the enrich logic should be in the tracer (`pkg/gadgets/trace/dns/tracer/tracer.go`) to avoid users of this package having to care about it.

Reply:
Another option is to define the callback right in the `Attach()` function and then we don't need any enrichment because we have the information about the container available there as done by @mqasimsarfraz in https://github.com/inspektor-gadget/inspektor-gadget/pull/1121. I'm wondering what is the cost of having many lamdba functions around.